src/kdf.rs

```diff
@@ -0,0 +1,351 @@
+// Copyright 2024 Simo Sorce
+// See LICENSE.txt file for terms
+
+use super::attribute;
+use super::err_rv;
+use super::error;
+use super::interface;
+use super::mechanism;
+use super::object;
+
+use attribute::from_bytes;
+use error::{KError, KResult};
+use interface::*;
+use mechanism::*;
+use object::{Object, ObjectFactories};
+
+use super::bytes_to_vec;
+
+use std::fmt::Debug;
+
+#[derive(Debug)]
+struct Sp800KDFMechanism {
+    info: CK_MECHANISM_INFO,
+    mech_type: CK_MECHANISM_TYPE,
+}
+
+impl Mechanism for Sp800KDFMechanism {
+    fn info(&self) -> &CK_MECHANISM_INFO {
+        &self.info
+    }
+
+    fn mech_type(&self) -> CK_MECHANISM_TYPE {
+        self.mech_type
+    }
+
+    fn derive_operation(&self, mech: &CK_MECHANISM) -> KResult<Operation> {
+        if self.info.flags & CKF_DERIVE != CKF_DERIVE {
+            return err_rv!(CKR_MECHANISM_INVALID);
+        }
+
+        let kdf = match mech.mechanism {
+            CKM_SP800_108_COUNTER_KDF => {
+                if mech.ulParameterLen as usize
+                    != ::std::mem::size_of::<CK_SP800_108_KDF_PARAMS>()
+                {
+                    return err_rv!(CKR_ARGUMENTS_BAD);
+                }
+                let kdf_params = unsafe {
+                    *(mech.pParameter as *const CK_SP800_108_KDF_PARAMS)
+                };
+                Sp800Operation::counter_kdf_new(kdf_params)?
+            }
+            CKM_SP800_108_FEEDBACK_KDF => {
+                if mech.ulParameterLen as usize
+                    != ::std::mem::size_of::<CK_SP800_108_FEEDBACK_KDF_PARAMS>()
+                {
+                    return err_rv!(CKR_ARGUMENTS_BAD);
+                }
+                let kdf_params = unsafe {
+                    *(mech.pParameter
+                        as *const CK_SP800_108_FEEDBACK_KDF_PARAMS)
+                };
+                Sp800Operation::feedback_kdf_new(kdf_params)?
+            }
+            CKM_SP800_108_DOUBLE_PIPELINE_KDF => {
```

Comment:
Ah, no double pipeline support, that's why i didn't see it in sp800_108. That's fine.

Reply:
Yeah OpenSSL simply does not have it, so I deferred adding it at all.